subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -179,26 +179,29 @@ private String getFilename(Path dir) {
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
             String name = internedToString(file.getFileName());
             if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
-                if (attrs == null) {
-                    throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
-                }
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
-                addFileSnapshot(file, name, attrs);
+                AbstractFileSystemLocationSnapshot snapshot = snapshot(file, attrs, name);
+                builder.visit(snapshot);
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private AbstractFileSystemLocationSnapshot snapshot(Path file, BasicFileAttributes attrs, String name) {
```

Comment:
This should return a `FileSystemLocationSnapshot`.

Reply:
Is it worth extracting this single-use method at all? The caller is a method with 5 lines of code altogether. By introducing this method we now have `snapshot()` and also `getFileSnapshot()` in the same source file. I'd inline this.

Reply:
* renamed to `getSnapshot`
* inlined `getFileSnapshot`

Reply:
I see `getSnapshot()` is back now. Will follow up at the later updates, closing this comment.